File: dataflow/graph.py
"""Graph registry with DAG validation and cycle detection."""

from collections import deque

from dataflow.primitives import Channel
from dataflow.node import Node


class GraphError(Exception):
    """Raised for graph construction errors (duplicates, cycles, etc.)."""


class Graph:
    """Directed acyclic graph of Nodes connected by Channels."""

    def __init__(self):
        self._nodes: dict[str, Node] = {}
        # Adjacency: src_name -> set of dst_names
        self._edges: dict[str, set[str]] = {}

    def add_node(self, node: Node):
        if node.name in self._nodes:
            raise GraphError(f"Duplicate node name: '{node.name}'")
        self._nodes[node.name] = node
        self._edges.setdefault(node.name, set())

    def connect(
        self,
        src: str,
        src_port: str,
        dst: str,
        dst_port: str,
        capacity: int = 64,
    ):
        """Create a Channel wiring src's output port to dst's input port."""
        if src not in self._nodes:
            raise GraphError(f"Unknown source node: '{src}'")
        if dst not in self._nodes:
            raise GraphError(f"Unknown destination node: '{dst}'")

        src_node = self._nodes[src]
        dst_node = self._nodes[dst]

        if src_port not in src_node.outputs:
            raise GraphError(f"Node '{src}' has no output port '{src_port}'")
        if dst_port not in dst_node.inputs:
            raise GraphError(f"Node '{dst}' has no input port '{dst_port}'")

        # Check for cycles: would adding src->dst create a cycle?
        if src == dst:
            raise GraphError(f"Self-loop on node '{src}'")
        if self._would_create_cycle(src, dst):
            raise GraphError(
                f"Adding edge {src}->{dst} would create a cycle"
            )

        channel = Channel(capacity)
        src_node.outputs[src_port].channel = channel
        dst_node.inputs[dst_port].channel = channel
        self._edges[src].add(dst)

    def _would_create_cycle(self, src: str, dst: str) -> bool:
        """Check if adding dst->...->src path exists (which means src->dst creates cycle)."""
        # If there's a path from dst to src, adding src->dst creates a cycle.
        visited = set()
        stack = [dst]
        while stack:
            node = stack.pop()
            if node == src:
                return True
            if node in visited:
                continue
            visited.add(node)
            stack.extend(self._edges.get(node, set()))
        return False

    def validate(self):
        """Check that all ports on all nodes are connected."""
        errors = []
        for name, node in self._nodes.items():
            for port_name, port in node.inputs.items():
                if not port.connected:
                    errors.append(f"Node '{name}' input '{port_name}' unconnected")
            for port_name, port in node.outputs.items():
                if not port.connected:
                    errors.append(f"Node '{name}' output '{port_name}' unconnected")
        if errors:
            raise GraphError("Validation failed:\n  " + "\n  ".join(errors))

    def source_nodes(self) -> list[Node]:
        """Nodes with no input ports."""
        return [n for n in self._nodes.values() if not n.inputs]

    def sink_nodes(self) -> list[Node]:
        """Nodes with no output ports."""
        return [n for n in self._nodes.values() if not n.outputs]

    def topological_order(self) -> list[Node]:
        """Kahn's algorithm for topological sort."""
        # Build in-degree map
        in_degree: dict[str, int] = {name: 0 for name in self._nodes}
        for src, dsts in self._edges.items():
            for dst in dsts:
                in_degree[dst] += 1

        queue = deque(name for name, deg in in_degree.items() if deg == 0)
        order = []

        while queue:
            name = queue.popleft()
            order.append(self._nodes[name])
            for dst in self._edges.get(name, set()):
                in_degree[dst] -= 1
                if in_degree[dst] == 0:
                    queue.append(dst)

        if len(order) != len(self._nodes):
            raise GraphError("Graph contains a cycle")

        return order

    @property
    def nodes(self) -> dict[str, Node]:
        return dict(self._nodes)

File: graphos-cpp/src/dataflow/graph.cpp
#include "graphos/dataflow/graph.hpp"
#include <deque>
#include <sstream>

namespace graphos {

void Graph::add_node(std::shared_ptr<Node> node) {
    const auto& name = node->name();
    if (nodes_.count(name))
        throw GraphError("Duplicate node name: '" + name + "'");
    edges_[name]; // init empty adjacency set
    nodes_.emplace(name, std::move(node));
}

void Graph::add_edge(const std::string& src, const std::string& dst) {
    edges_[src].insert(dst);
    if (edges_.find(dst) == edges_.end())
        edges_[dst]; // ensure dst has an entry
}

bool Graph::would_create_cycle(const std::string& src,
                               const std::string& dst) const {
    // DFS from dst — if we can reach src, adding src→dst creates a cycle
    std::unordered_set<std::string> visited;
    std::vector<std::string> stack = {dst};

    while (!stack.empty()) {
        auto node = std::move(stack.back());
        stack.pop_back();
        if (node == src) return true;
        if (visited.count(node)) continue;
        visited.insert(node);
        auto it = edges_.find(node);
        if (it != edges_.end()) {
            for (const auto& neighbor : it->second) {
                stack.push_back(neighbor);
            }
        }
    }
    return false;
}

void Graph::validate() const {
    // For now, just ensure graph is non-empty and acyclic (topo sort will catch cycles)
    if (nodes_.empty())
        throw GraphError("Graph has no nodes");
    // Try topo sort — throws on cycle
    topological_order();
}

std::vector<std::shared_ptr<Node>> Graph::topological_order() const {
    // Kahn's algorithm
    std::unordered_map<std::string, int> in_degree;
    for (const auto& [name, _] : nodes_) {
        in_degree[name] = 0;
    }
    for (const auto& [src, dsts] : edges_) {
        for (const auto& dst : dsts) {
            if (in_degree.count(dst))
                in_degree[dst]++;
        }
    }

    std::deque<std::string> queue;
    for (const auto& [name, deg] : in_degree) {
        if (deg == 0) queue.push_back(name);
    }

    std::vector<std::shared_ptr<Node>> order;
    order.reserve(nodes_.size());

    while (!queue.empty()) {
        auto name = std::move(queue.front());
        queue.pop_front();
        order.push_back(nodes_.at(name));
        auto it = edges_.find(name);
        if (it != edges_.end()) {
            for (const auto& dst : it->second) {
                if (--in_degree[dst] == 0) {
                    queue.push_back(dst);
                }
            }
        }
    }

    if (order.size() != nodes_.size())
        throw GraphError("Graph contains a cycle");

    return order;
}

void Graph::close_all_channels() noexcept {
    for (auto& closer : channels_) {
        try { closer(); } catch (...) {}
    }
}

} // namespace graphos
